platform/src/main/java/org/hillview/remoting/HillviewServer.java

```diff
@@ -137,32 +171,50 @@ public void flatMap(final Command command, final StreamObserver<PartialResponse>
      */
     @Override
     @SuppressWarnings("unchecked")
-    public void sketch(final Command command,
-                       final StreamObserver<PartialResponse> responseObserver) {
+    public void sketch(final Command command, final StreamObserver<PartialResponse> responseObserver) {
         try {
             if (!this.checkValidIdsIndex(command.getIdsIndex(), responseObserver)) {
                 return;
             }
+            if (this.respondIfReplyIsMemoized(command, responseObserver)) {
+                LOG.info("Returning memoized result for sketch operation against IDataSet#" + command.getIdsIndex());
+                return;
+            }
             final byte[] bytes = command.getSerializedOp().toByteArray();
             final SketchOperation sketchOp = SerializationUtils.deserialize(bytes);
             final Observable<PartialResult> observable = this.dataSets.get(command.getIdsIndex())
                                                                       .sketch(sketchOp.sketch);
+            final UUID commandId = new UUID(command.getHighId(), command.getLowId());
             final Subscription sub = observable.subscribe(new Subscriber<PartialResult>() {
+                @Nullable private Object result = sketchOp.sketch.getZero();
```

Comment:
Let's call this one sketchResultAccumulator.
Also, use this. for referring to it.